File: bindings/csharp/src/STRling/Core/Nodes.cs
using System.Text.Json.Serialization;
using System.Collections.Generic;

namespace Strling.Core;

public record Flags
{
    public bool IgnoreCase { get; init; } = false;
    public bool Multiline { get; init; } = false;
    public bool DotAll { get; init; } = false;
    public bool Unicode { get; init; } = false;
    public bool Extended { get; init; } = false;

    public static Flags FromLetters(string letters)
    {
        var f = new Flags();
        foreach (var c in letters)
        {
            if (c == 'i') f = f with { IgnoreCase = true };
            else if (c == 'm') f = f with { Multiline = true };
            else if (c == 's') f = f with { DotAll = true };
            else if (c == 'u') f = f with { Unicode = true };
            else if (c == 'x') f = f with { Extended = true };
        }
        return f;
    }
}

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(Alt), typeDiscriminator: "Alternation")]
[JsonDerivedType(typeof(Seq), typeDiscriminator: "Sequence")]
[JsonDerivedType(typeof(Lit), typeDiscriminator: "Literal")]
[JsonDerivedType(typeof(Dot), typeDiscriminator: "Dot")]
[JsonDerivedType(typeof(Anchor), typeDiscriminator: "Anchor")]
[JsonDerivedType(typeof(CharClass), typeDiscriminator: "CharacterClass")]
[JsonDerivedType(typeof(Quant), typeDiscriminator: "Quantifier")]
[JsonDerivedType(typeof(Group), typeDiscriminator: "Group")]
[JsonDerivedType(typeof(Backref), typeDiscriminator: "Backreference")]
[JsonDerivedType(typeof(Lookahead), typeDiscriminator: "Lookahead")]
[JsonDerivedType(typeof(NegativeLookahead), typeDiscriminator: "NegativeLookahead")]
[JsonDerivedType(typeof(Lookbehind), typeDiscriminator: "Lookbehind")]
[JsonDerivedType(typeof(NegativeLookbehind), typeDiscriminator: "NegativeLookbehind")]
public abstract record Node;

public record Alt([property: JsonPropertyName("alternatives")] List<Node> Alternatives) : Node;

public record Seq([property: JsonPropertyName("parts")] List<Node> Parts) : Node;

public record Lit([property: JsonPropertyName("value")] string Value) : Node;

public record Dot : Node;

public record Anchor([property: JsonPropertyName("at")] string At) : Node;

public record CharClass(
    [property: JsonPropertyName("negated")] bool Negated,
    [property: JsonPropertyName("members")] List<ClassItem> Members
) : Node;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(ClassRange), typeDiscriminator: "Range")]
[JsonDerivedType(typeof(ClassLiteral), typeDiscriminator: "Literal")]
[JsonDerivedType(typeof(ClassEscape), typeDiscriminator: "Escape")]
[JsonDerivedType(typeof(ClassUnicodeProperty), typeDiscriminator: "UnicodeProperty")]
public abstract record ClassItem;

public record ClassRange(
    [property: JsonPropertyName("from")] string From,
    [property: JsonPropertyName("to")] string To
) : ClassItem;

public record ClassLiteral(
    [property: JsonPropertyName("value")] string Value
) : ClassItem;

public record ClassEscape(
    [property: JsonPropertyName("kind")] string Kind
) : ClassItem;

public record ClassUnicodeProperty(
    [property: JsonPropertyName("name")] string? Name,
    [property: JsonPropertyName("value")] string Value,
    [property: JsonPropertyName("negated")] bool Negated
) : ClassItem;

public record Quant(
    [property: JsonPropertyName("target")] Node Target,
    [property: JsonPropertyName("min")] int Min,
    [property: JsonPropertyName("max")] int? Max,
    [property: JsonPropertyName("greedy")] bool Greedy,
    [property: JsonPropertyName("lazy")] bool Lazy,
    [property: JsonPropertyName("possessive")] bool Possessive
) : Node;

public record Group(
    [property: JsonPropertyName("capturing")] bool Capturing,
    [property: JsonPropertyName("body")] Node Body,
    [property: JsonPropertyName("name")] string? Name,
    [property: JsonPropertyName("atomic")] bool Atomic
) : Node;

public record Backref(
    [property: JsonPropertyName("index")] int? Index,
    [property: JsonPropertyName("name")] string? Name
) : Node;

public record Lookahead([property: JsonPropertyName("body")] Node Body) : Node;
public record NegativeLookahead([property: JsonPropertyName("body")] Node Body) : Node;
public record Lookbehind([property: JsonPropertyName("body")] Node Body) : Node;
public record NegativeLookbehind([property: JsonPropertyName("body")] Node Body) : Node;

File: bindings/kotlin/src/main/kotlin/strling/core/Nodes.kt
package strling.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import kotlinx.serialization.json.JsonElement

/**
 * STRling AST Node Definitions
 * 
 * This module defines the complete set of Abstract Syntax Tree (AST) node classes
 * that represent the parsed structure of STRling patterns.
 */

// ---- Flags container ----

/**
 * Container for regex flags/modifiers.
 */
@Serializable
data class Flags(
    val ignoreCase: Boolean = false,
    val multiline: Boolean = false,
    val dotAll: Boolean = false,
    val unicode: Boolean = false,
    val extended: Boolean = false
) {
    companion object {
        /**
         * Creates Flags from a string of flag letters.
         */
        fun fromLetters(letters: String): Flags {
            var ignoreCase = false
            var multiline = false
            var dotAll = false
            var unicode = false
            var extended = false
            
            val cleaned = letters.replace(",", "").replace(" ", "")
            for (ch in cleaned) {
                when (ch) {
                    'i' -> ignoreCase = true
                    'm' -> multiline = true
                    's' -> dotAll = true
                    'u' -> unicode = true
                    'x' -> extended = true
                }
            }
            
            return Flags(ignoreCase, multiline, dotAll, unicode, extended)
        }
    }
}

// ---- Base node ----

@Serializable
@JsonClassDiscriminator("type")
sealed interface Node

@Serializable
@JsonClassDiscriminator("type")
sealed interface ClassItem

// ---- Concrete nodes ----

@Serializable
@SerialName("Alternation")
data class Alternation(val alternatives: List<Node>) : Node

@Serializable
@SerialName("Sequence")
data class Sequence(val parts: List<Node>) : Node

@Serializable
@SerialName("Literal")
data class Literal(val value: String) : Node, ClassItem

@Serializable
@SerialName("Dot")
class Dot : Node {
    override fun equals(other: Any?): Boolean = other is Dot
    override fun hashCode(): Int = "Dot".hashCode()
}

@Serializable
@SerialName("Anchor")
data class Anchor(val at: String) : Node

@Serializable
@SerialName("CharacterClass")
data class CharacterClass(val negated: Boolean, val members: List<ClassItem>) : Node

@Serializable
@SerialName("Range")
data class Range(val from: String, val to: String) : ClassItem

@Serializable
@SerialName("Escape")
data class Escape(val kind: String) : ClassItem

@Serializable
@SerialName("UnicodeProperty")
data class UnicodeProperty(
    val value: String,
    val name: String? = null,
    val negated: Boolean = false
) : ClassItem

@Serializable
@SerialName("Quantifier")
data class Quantifier(
    val target: Node,
    val min: Int,
    val max: JsonElement? = null,
    val greedy: Boolean,
    val lazy: Boolean,
    val possessive: Boolean
) : Node

@Serializable
@SerialName("Group")
data class Group(
    val capturing: Boolean,
    val body: Node,
    val name: String? = null,
    val atomic: Boolean? = null,
    val number: Int? = null,
    val expression: Node? = null
) : Node

@Serializable
@SerialName("Backreference")
data class Backreference(
    val index: Int? = null,
    val name: String? = null
) : Node

@Serializable
@SerialName("Lookahead")
data class Lookahead(val body: Node) : Node

@Serializable
@SerialName("NegativeLookahead")
data class NegativeLookahead(val body: Node) : Node

@Serializable
@SerialName("Lookbehind")
data class Lookbehind(val body: Node) : Node

@Serializable
@SerialName("NegativeLookbehind")
data class NegativeLookbehind(val body: Node) : Node
